### PR TITLE
Tempus: Remove ParameterList from SolutionHistory

### PR DESCRIPTION
**Pull request by ccober6**
* SolutionHistory is no longer inherits from ParameterListAcceptor.
 * Moved SolutionHistory constructor with ParameterList to a
   nonmember constructor, createSolutionHistoryPL(pl).
 * Added function to set the history, setHistory().
 * Simplified the internal management of "Storage Type".
   - Can set with enum StorageType or a string.
   - Removed ParameterList validator for enum storageType.
 * Added methods to initialize SolutionHistory similar to
   other Tempus objects.
 * Added checks to addState to ensure new state's index does not
   match an index already in the SolutionHistory. This can
   be turned off.
 * Fixed bug in addWorkingState, where state was added to
   history (chronologically) and then the TIME was changed.
 * Added unit tests.

@trilinos/tempus

### CODE REVIEW

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_SolutionHistory_decl.hpp:249]: please add space and doxygen to these

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_SolutionHistory_decl.hpp:252]: same here

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_SolutionHistory_decl.hpp:286]: here too

File: packages/tempus/test/Integrator/Tempus_IntegratorBasic_ref.xml
Comment (sconde) [inline packages/tempus/test/Integrator/Tempus_IntegratorBasic_ref.xml:11]: The order is now required?

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (ccober6) [reply packages/tempus/src/Tempus_SolutionHistory_decl.hpp:249]: Done!

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (ccober6) [reply packages/tempus/src/Tempus_SolutionHistory_decl.hpp:252]: Done!

File: packages/tempus/test/Integrator/Tempus_IntegratorBasic_ref.xml
Comment (ccober6) [reply packages/tempus/test/Integrator/Tempus_IntegratorBasic_ref.xml:11]: No, but to check if the defaults have been set correctly, I needed to have it set in the reference ParameterList.

**Comment by ccober6** [comment]
@sconde Do you have anything else I should look at?  Otherwise I will commit the previous changes and get the final review approval.  :)

**Comment by sconde** [comment]
Nothing more. When you push the changes, I will approve the PR

**Comment by ccober6** [comment]
@sconde FYI, I just pushed those changes, so you can approve again when you are ready.  Thanks!

**Comment by ccober6** [comment]
@sconde Just a quick reminder on approving this.  Thanks, Curt

**Comment by ccober6** [comment]
@rppawlo Can you also review this one?  Thanks!

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (rppawlo) [inline packages/tempus/src/Tempus_SolutionHistory_decl.hpp:327]: Should this be a const function?

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (rppawlo) [inline packages/tempus/src/Tempus_SolutionHistory_decl.hpp:340]: Why do you need mutable? Shouldn't the state of this flag should only change in non-const functions that change the state of the object?

**Review by rppawlo — APPROVED** [review]

File: packages/tempus/src/Tempus_SolutionHistory_decl.hpp
Comment (ccober6) [reply packages/tempus/src/Tempus_SolutionHistory_decl.hpp:340]: So here was the thought.  Please let me know if you have a better way to do this.

The initialize() function is const, so that the object is not changed during initialize() (e.g., no expensive operations are done: memory allocation, etc.).  However we need to set isInitialized_ = true, if everything is initialized.  Thus I made isInitialized_ mutable.  Thoughts?

**Pull request merged**
